File: platforms/android/keyboard-ui/src/main/java/app/funput/funput/keyboard/ui/emoji/EmojiSearchIndex.kt
package app.funput.funput.keyboard.ui.emoji

import java.text.Normalizer
import java.util.Locale

internal class EmojiSearchIndex(items: List<EmojiItem>) {
    private val entries = items.mapIndexed { index, item -> Entry(index, item) }

    fun search(query: String, limit: Int = 64): List<EmojiItem> {
        val words = normalize(query).split(' ').filter(String::isNotBlank)
        if (words.isEmpty()) return emptyList()
        return entries.mapNotNull { entry ->
            rank(entry, words)?.let { rank -> Ranked(rank, entry.index, entry.item) }
        }.sortedWith(compareBy<Ranked> { it.rank }.thenBy { it.index })
            .take(limit)
            .map(Ranked::item)
    }

    private fun rank(entry: Entry, query: List<String>): Int? {
        val phrase = query.joinToString(" ")
        if (entry.names.any { it == phrase }) return 0
        if (entry.names.any { it.startsWith(phrase) }) return 1
        if (query.all { word -> entry.words.any { it.startsWith(word) } }) return 2
        if (query.all { word -> entry.terms.any { it.contains(word) } }) return 3
        return null
    }

    private data class Entry(val index: Int, val item: EmojiItem) {
        val names = listOfNotNull(item.localizedName, item.name).map(::normalize)
        val terms = (item.searchTerms + names).map(::normalize)
        val words = terms.flatMap { it.split(' ') }.filter(String::isNotBlank)
    }

    private data class Ranked(val rank: Int, val index: Int, val item: EmojiItem)

    companion object {
        internal fun normalize(value: String): String = Normalizer
            .normalize(value.lowercase(Locale.ROOT).replace('đ', 'd'), Normalizer.Form.NFKD)
            .replace(Regex("\\p{M}+"), "")
            .replace(Regex("[^\\p{L}\\p{N}]+"), " ")
            .trim()
            .replace(Regex("\\s+"), " ")
    }
}

File: platforms/ios/Packages/FunputKit/Sources/KeyboardRenderer/Emoji/Catalog/EmojiSearchIndex.swift
import Foundation

struct EmojiSearchIndex {
    private struct Entry {
        let item: EmojiItem
        let names: [String]
        let terms: [String]
        let words: [String]
    }

    private let entries: [Entry]

    init(catalog: EmojiCatalog) {
        entries = catalog.emojis.map { item in
            let names = [item.localizedName, item.name]
                .compactMap { $0 }
                .map(Self.normalize)
            let terms = (names + item.searchTerms.map(Self.normalize))
                .filter { !$0.isEmpty }
            return Entry(
                item: item,
                names: names,
                terms: terms,
                words: terms.flatMap { $0.split(separator: " ").map(String.init) }
            )
        }
    }

    func search(_ query: String, limit: Int = 64) -> [EmojiItem] {
        let normalized = Self.normalize(query)
        guard !normalized.isEmpty, limit > 0 else { return [] }
        let queryWords = normalized.split(separator: " ").map(String.init)
        var buckets = Array(repeating: [EmojiItem](), count: 4)

        for entry in entries {
            guard let score = score(entry, query: normalized, words: queryWords) else { continue }
            buckets[score].append(entry.item)
        }
        return buckets.flatMap { $0 }.prefix(limit).map { $0 }
    }

    private func score(_ entry: Entry, query: String, words: [String]) -> Int? {
        if entry.names.contains(query) { return 0 }
        if entry.names.contains(where: { $0.hasPrefix(query) }) { return 1 }
        if words.allSatisfy({ word in entry.words.contains(where: { $0.hasPrefix(word) }) }) {
            return 2
        }
        if words.allSatisfy({ word in entry.terms.contains(where: { $0.contains(word) }) }) {
            return 3
        }
        return nil
    }

    static func normalize(_ value: String) -> String {
        let folded = value.folding(
            options: [.caseInsensitive, .diacriticInsensitive, .widthInsensitive],
            locale: Locale(identifier: "vi_VN")
        )
        let words = folded.unicodeScalars.split { scalar in
            !CharacterSet.alphanumerics.contains(scalar)
        }
        return words.map(String.init).joined(separator: " ")
    }
}
